File: ios/services/ConfigurationBuilder.swift
//
//  ConfigurationBuilder.swift
//  LunarDatePicker
//
//  Service for building picker configurations
//

import Foundation

final class ConfigurationBuilder: ConfigurationProviding {

  // MARK: - ConfigurationProviding

  func buildPickerConfig(
    from params: LDP_PresentParams,
    globalConfig: LDP_ConfigParams?
  ) -> PickerConfig {
    var pickerConfig = PickerConfig.default

    // Apply basic configuration
    pickerConfig.controller.title = params.title
    pickerConfig.controller.isSingleMode = (params.mode == LDP_Mode.single)

    // Apply theme if available
    if let globalConfig = globalConfig,
      let theme = globalConfig.themes[params.theme]
    {
      applyTheme(theme, to: &pickerConfig)
    }

    // Apply language if available
    if let globalConfig = globalConfig,
      let language = globalConfig.languages[params.language]
    {
      applyLanguage(language, to: &pickerConfig)
    }

    // Apply timezone if available
    if let timeZoneOffset = globalConfig?.timeZoneOffset {
      applyTimeZone(timeZoneOffset, to: &pickerConfig)
    }

    // Apply year range if available
    if let yearRangeOffset = globalConfig?.yearRangeOffset {
      applyYearRange(yearRangeOffset, to: &pickerConfig)
    }

    // Apply global settings if available in globalConfig
    if let globalConfig = globalConfig {
      pickerConfig.controller.showSubmitButton = globalConfig.showSubmitButton
    }

    // Force showSubmitButton to false in single mode
    if params.mode == .single {
      pickerConfig.controller.showSubmitButton = false
    }

    return pickerConfig
  }

  func applyTheme(_ theme: LDP_CustomStyle, to config: inout PickerConfig) {
    // Controller colors
    config.controller.backgroundColor = ColorWrapper(
      fromHex: theme.backgroundColor
    )
    config.controller.titleColor = ColorWrapper(fromHex: theme.titleColor)
    config.controller.secondaryTextColor = ColorWrapper(fromHex: theme.secondColor)
    config.controller.submitButtonColor = ColorWrapper(fromHex: theme.submitButtonColor)

    // Day cell colors
    config.dayCell.dateLabelColor = ColorWrapper(fromHex: theme.dateLabelColor)
    config.dayCell.todayLabelColor = ColorWrapper(fromHex: theme.todayLabelColor)
    config.dayCell.weekendLabelColor = ColorWrapper(
      fromHex: theme.weekendLabelColor
    )
    config.dayCell.lunarDateLabelColor = ColorWrapper(
      fromHex: theme.lunarDateLabelColor
    )
    config.dayCell.specialDateLabelColor = ColorWrapper(
      fromHex: theme.specialDayLabelColor
    )
    config.dayCell.rangeBackgroundColor = ColorWrapper(
      fromHex: theme.rangeBackgroundColor
    )
    config.dayCell.selectedBackgroundColor = ColorWrapper(
      fromHex: theme.selectedBackgroundColor
    )
    config.dayCell.selectedTextColor = ColorWrapper(
      fromHex: theme.selectedTextColor
    )

    // Month header colors
    config.monthHeader.labelColor = ColorWrapper(fromHex: theme.monthLabelColor)

    // Week view colors
    config.weekView.backgroundColor = ColorWrapper(
      fromHex: theme.weekViewBackgroundColor
    )
    config.weekView.weekendLabelColor = ColorWrapper(
      fromHex: theme.weekendLabelColor
    )
    config.weekView.weekLabelColor = ColorWrapper(fromHex: theme.dateLabelColor)
  }

  func applyLanguage(_ language: LDP_CustomLanguage, to config: inout PickerConfig)
  {
    config.weekView.weekdayNames = language.weekdayNames
    // Apply locale for all native date formatting
    config.calendar.locale = Locale(identifier: language.locale)
  }

  func applyTimeZone(_ offset: Double, to config: inout PickerConfig) {
    config.calendar.timeZone =
      TimeZone(secondsFromGMT: Int(offset * 3600)) ?? TimeZone.current
  }

  func applyYearRange(_ offset: Double, to config: inout PickerConfig) {
    config.yearRangeOffset = Int(offset)
  }
}

File: android/src/main/java/com/margelo/nitro/lunardatepicker/services/ConfigurationBuilder.kt
package com.margelo.nitro.lunardatepicker.services

import com.margelo.nitro.lunardatepicker.*
import com.margelo.nitro.lunardatepicker.models.*
import com.margelo.nitro.lunardatepicker.utils.*
import com.margelo.nitro.lunardatepicker.constants.DataConstants
import android.util.Log

import java.util.*

class ConfigurationBuilder {

  companion object {
    private const val TAG = DataConstants.LogTags.CONFIG_BUILDER
  }

  private val cacheService: ConfigurationCacheService by lazy {
    ConfigurationCacheService()
  }

  fun buildPickerConfig(params: LDP_PresentParams, globalConfig: LDP_ConfigParams?): PickerConfig {
    return try {
      cacheService.getOrBuildConfiguration(params, globalConfig)
    } catch (e: Exception) {
      Log.e(TAG, "Failed to build configuration using cache service, falling back to manual build", e)
      buildConfigurationManually(params, globalConfig)
    }
  }

  private fun buildConfigurationManually(params: LDP_PresentParams, globalConfig: LDP_ConfigParams?): PickerConfig {
    var pickerConfig = PickerConfig.default.copy()

    pickerConfig = pickerConfig.copy(
      controller = pickerConfig.controller.copy(
        title = params.title,
        isSingleMode = params.mode == LDP_Mode.SINGLE
      )
    )

    globalConfig?.let { config ->
      config.themes[params.theme]?.let { theme ->
        pickerConfig = applyTheme(theme, pickerConfig)
      }
      config.languages[params.language]?.let { language ->
        pickerConfig = applyLanguage(language, pickerConfig)
      }
      pickerConfig = applyTimeZone(config.timeZoneOffset, pickerConfig)
      pickerConfig = applyYearRange(config.yearRangeOffset, pickerConfig)
      // Map showSubmitButton from global config
      pickerConfig = pickerConfig.copy(
        controller = pickerConfig.controller.copy(
          showSubmitButton = config.showSubmitButton
        )
      )
    }

    // Force showSubmitButton to false in single mode
    if (params.mode == LDP_Mode.SINGLE) {
      pickerConfig = pickerConfig.copy(
        controller = pickerConfig.controller.copy(
          showSubmitButton = false
        )
      )
    }

    return pickerConfig
  }

  private fun applyTheme(theme: LDP_CustomStyle, config: PickerConfig): PickerConfig {
    return config.copy(
      controller = config.controller.copy(
        backgroundColor = ColorUtils.colorFromHex(theme.backgroundColor),
        titleColor = ColorUtils.colorFromHex(theme.titleColor),
        secondaryTextColor = ColorUtils.colorFromHex(theme.secondColor),
        submitButtonColor = ColorUtils.colorFromHex(theme.submitButtonColor),
        noticeLabelColor = ColorUtils.colorFromHex(theme.noticeLabelColor),
        noticeBackgroundColor = ColorUtils.colorFromHex(theme.noticeBackgroundColor)
      ),
      dayCell = config.dayCell.copy(
        dateLabelColor = ColorUtils.colorFromHex(theme.dateLabelColor),
        todayLabelColor = ColorUtils.colorFromHex(theme.todayLabelColor),
        weekendLabelColor = ColorUtils.colorFromHex(theme.weekendLabelColor),
        lunarDateLabelColor = ColorUtils.colorFromHex(theme.lunarDateLabelColor),
        specialDateLabelColor = ColorUtils.colorFromHex(theme.specialDayLabelColor),
        priceLabelColor = ColorUtils.colorFromHex(theme.priceLabelColor),
        cheapestPriceLabelColor = ColorUtils.colorFromHex(theme.cheapestPriceLabelColor),
        rangeBackgroundColor = ColorUtils.colorFromHex(theme.rangeBackgroundColor),
        selectedBackgroundColor = ColorUtils.colorFromHex(theme.selectedBackgroundColor),
        selectedTextColor = ColorUtils.colorFromHex(theme.selectedTextColor)
      ),
      monthHeader = config.monthHeader.copy(
        labelColor = ColorUtils.colorFromHex(theme.monthLabelColor)
      ),
      weekView = config.weekView.copy(
        backgroundColor = ColorUtils.colorFromHex(theme.weekViewBackgroundColor),
        weekendLabelColor = ColorUtils.colorFromHex(theme.weekendLabelColor),
        weekLabelColor = ColorUtils.colorFromHex(theme.dateLabelColor)
      )
    )
  }

  private fun applyLanguage(language: LDP_CustomLanguage, config: PickerConfig): PickerConfig {
    return config.copy(
      weekView = config.weekView.copy(
        weekdayNames = language.weekdayNames
      ),
      calendar = config.calendar.copy(
        locale = Locale(language.locale)
      )
    )
  }

  private fun applyTimeZone(offset: Double, config: PickerConfig): PickerConfig {
    val timeZone = TimeZone.getTimeZone("GMT${if (offset >= 0) "+" else ""}${offset.toInt()}")
    return config.copy(
      calendar = config.calendar.copy(
        timeZone = timeZone
      )
    )
  }

  private fun applyYearRange(offset: Double, config: PickerConfig): PickerConfig {
    return config.copy(
      yearRangeOffset = offset.toInt()
    )
  }
}